### ISSUE TITLE
[FEAT] Allow to use existing app service plan

### ISSUE THREAD

**Issue by kevin-fitlg** [claranet/terraform-azurerm-function-app#3]
### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Hi team
App service plans are generally consumed by multiple function apps (or app services);
This module expects to have a one-to-one mapping between function app and underlying app service plan, by enforcing creation of a new app-service-plan.
Can you explain why are you enforcing this? and if you are open to make this conditional, i.e. when `var.service_plan_id` is provided, the module does not create any new app service plan, but rather consume the existing one.
Similar suggestion is valid for app-service module.

Note that we assume the user knows and understand which app-service-plan is suitable for the new function-app and therefore we don't need to perform any extra checks on the provided `var.service_plan_id`.

Thanks

### New or Affected Resource(s)/Data Source(s)

azurerm_app_service_plan

### Potential Terraform Configuration

```hcl
### Windows
module "function_app_windows" {
  source  = "claranet/function-app/azurerm"
  version = "x.x.x"

  client_name         = var.client_name
  environment         = var.environment
  stack               = var.stack
  location            = module.azure_region.location
  location_short      = module.azure_region.location_short
  resource_group_name = module.rg.resource_group_name

  name_prefix = "hello"

  service_plan_id = "/subscription/x.y.z"

  function_app_application_settings = {
    "tracker_id"      = "AJKGDFJKHFDS"
    "backend_api_url" = "https://backend.domain.tld/api"
  }

  storage_account_identity_type = "SystemAssigned"

  logs_destinations_ids = [
    module.logs.logs_storage_account_id,
    module.logs.log_analytics_workspace_id
  ]

  extra_tags = {
    foo = "bar"
  }
}

### Linux
module "function_app_linux" {
  source  = "claranet/function-app/azurerm"
  version = "x.x.x"

  client_name         = var.client_name
  environment         = var.environment
  stack               = var.stack
  location            = module.azure_region.location
  location_short      = module.azure_region.location_short
  resource_group_name = module.rg.resource_group_name

  name_prefix = "hello"

  service_plan_id = "/subscription/x.y.z"

  function_app_version = 4
  function_app_site_config = {
    application_stack = {
      python_version = "3.9"
    }
  }

  function_app_application_settings = {
    "tracker_id"      = "AJKGDFJKHFDS"
    "backend_api_url" = "https://backend.domain.tld/api"
  }

  storage_account_identity_type = "SystemAssigned"

  logs_destinations_ids = [
    module.logs.logs_storage_account_id,
    module.logs.log_analytics_workspace_id
  ]

  extra_tags = {
    foo = "bar"
  }
}
```

**Comment by Shr3ps** [comment]
Hello,

I think that using the sub-modules should be enough for this. What do you think?
ref: https://github.com/claranet/terraform-azurerm-function-app/blob/master/modules/linux-function/README.md

**Comment by kevin-fitlg** [comment]
Thanks for point to the submodule, I totally forgot that I can refer to them.
So, I tried it @Shr3ps and I noticed that in the submodule you have a [data source block](https://github.com/claranet/terraform-azurerm-function-app/blob/02f2d575b5187d7351440ffdc975502c3847ede7/modules/windows-function/r-function.tf#L2-L5) like this:
```
# Data Service Plan
data "azurerm_service_plan" "plan" {
  name                = element(split("/", var.service_plan_id), 8)
  resource_group_name = var.resource_group_name
}
```
This data source fails if the app service plan is inside a different resource group (separate from function apps resource group).
As I searched, this data source is not referenced anywhere. So, is it just to make sure the app service plan exists?
If that is the case, which I think is not necessary to have a data source, we can extract it from the `service_plan_id` ; Something like:
```
# Data Service Plan
data "azurerm_service_plan" "plan" {
  name                = element(split("/", var.service_plan_id), 8)
  resource_group_name = element(split("/", var.service_plan_id), 4)
}
```

**Comment by kevin-fitlg** [comment]
Further investigated this, and I realized this module heavily relies on using storage accounts to store ZIP packages for each function app.
I don't need any of those storage account components to be created really.
That said, even when I try to just create a single storage account and use the option `var.use_existing_storage_account` for bypass creation of extra components, the module wants to create same blob container "functions-packages" for all functions in that single storage account, which I expect to fail (because the container already created for the first function app).
I think I am going to skip using this module, it is not generic enough, unless you can show me some magical solution.

**Comment by BzSpi** [comment]
Hi @kevin-fitlg 

The module does not heavily rely on storage account for zip files, this is an additional feature only relevant in specific use cases we decided to implement (since your application should not be linked to your infrastructure deployment) as the one [here](https://github.com/claranet/terraform-azurerm-run/blob/master/modules/monitoring-function/r-function.tf#L38).
The container is created only if you use the "zip deploy" feature from this module.
Also, this feature has been implemented in Terraform a couple of weeks ago.

A function-app service **always needs** a storage account no matter your use case: https://learn.microsoft.com/en-us/azure/azure-functions/storage-considerations?tabs=azure-cli#storage-account-requirements
